fxa-rust-client/src/scoped_keys.rs

```diff
@@ -0,0 +1,101 @@
+use errors::*;
+
+use base64;
+use byteorder::{ByteOrder, BigEndian};
+use ring;
+use ring::{aead, agreement, rand};
+use ring::agreement::EphemeralPrivateKey;
+use serde_json;
+use sha2::{Sha256, Digest};
+use untrusted::Input;
+
+pub struct ScopedKeysFlow {
+  private_key: EphemeralPrivateKey
+}
+
+impl ScopedKeysFlow {
+  pub fn with_random_ecdh_key() -> Result<ScopedKeysFlow> {
+    let rng = rand::SystemRandom::new();
+    let private_key = EphemeralPrivateKey::generate(&agreement::ECDH_P256, &rng)?;
+    Ok(ScopedKeysFlow {
+      private_key
+    })
+  }
+
+  pub fn generate_keys_jwk(&self) -> Result<String> {
+    let mut pub_key = vec![0u8; self.private_key.public_key_len()];
+    &self.private_key.compute_public_key(&mut pub_key)?;
+    // First byte is 4, then 32 bytes for x, and 32 bytes for y.
+    assert_eq!(pub_key.len(), 1 + 32 + 32);
+    assert_eq!(pub_key[0], 0x04);
+    let x = Vec::from(&pub_key[1..33]);
+    let x = base64::encode_config(&x, base64::URL_SAFE_NO_PAD);
+    let y = Vec::from(&pub_key[33..]);
+    let y = base64::encode_config(&y, base64::URL_SAFE_NO_PAD);
+    Ok(json!({
+        "crv": "P-256",
+        "kty": "EC",
+        "x": x,
+        "y": y,
+    }).to_string())
+  }
+
+  pub fn decrypt_keys_jwe(self, jwe: &str) -> Result<String> {
+    let segments: Vec<&str> = jwe.split(".").collect();
+    let header = base64::decode_config(&segments[0], base64::URL_SAFE_NO_PAD)?;
+    let protected_header: serde_json::Value = serde_json::from_slice(&header)?;
+    assert_eq!(protected_header["epk"]["kty"], "EC");
+    assert_eq!(protected_header["epk"]["crv"], "P-256");
+
+    // Part 1: Grab the x/y from the other party and construct the secret.
+    let x = base64::decode_config(&protected_header["epk"]["x"].as_str().unwrap(), base64::URL_SAFE_NO_PAD)?;
+    let y = base64::decode_config(&protected_header["epk"]["y"].as_str().unwrap(), base64::URL_SAFE_NO_PAD)?;
+    let mut peer_pub_key: Vec<u8> = vec![0x04];
+    peer_pub_key.extend_from_slice(&x);
+    peer_pub_key.extend_from_slice(&y);
+    let peer_pub_key = Input::from(&peer_pub_key);
+    let secret = agreement::agree_ephemeral(self.private_key, &agreement::ECDH_P256, peer_pub_key, ring::error::Unspecified, |z| {
+        // ConcatKDF
+        let counter = 1;
+        let alg = protected_header["enc"].as_str().unwrap();
+        let apu = "";
+        let apv = "";
+        let mut buf: Vec<u8> = vec![];
+        buf.extend_from_slice(&to_32b_buf(counter));
+        buf.extend_from_slice(&z);
+        buf.extend_from_slice(&to_32b_buf(alg.len() as u32));
```

Comment:
maybe include a comment this is ConcatKDF's "otherinfo" -- again just thinking about future us.